src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Poppify.cpp - Transform Binaryen IR to Poppy IR.
+//
+// Poppy IR represents stack machine code using normal Binaryen IR types by
+// imposing the following constraints:
+//
+//  1. Function bodies and children of control flow (except If conditions) must
+//     be blocks.
+//
+//  2. Blocks may have any Expressions except for Pops as children. The sequence
+//     of instructions in a block follows the same validation rules as in
+//     WebAssembly. That means that any expression may have a concrete type, not
+//     just the final expression in the block.
+//
+//  3. All other children must be Pops, which are used to determine the input
+//     stack type of each instruction. Pops may not have `unreachable` type.
+//
+//  4. Only control flow structures and instructions that have polymorphic
+//     unreachable behavior in WebAssembly may have unreachable type. Blocks may
+//     be unreachable when they are not branch targets and when they have an
+//     unreachable child. Note that this means a block may be unreachable even
+//     if it would otherwise have a concrete type, unlike in Binaryen IR. For
+//     example, this block could have unreachable type in Poppy IR but would
+//     have to have type i32 in Binaryen IR:
```

Comment:
Thanks, now I understand the difference.

But why is this so? It's not clear to me why it has to be this way. (If it doesn't, removing differences would be good - but I assume there is a reason...)

Reply:
TBH I don't remember the original concrete motivation for making this change in Poppy IR. Allowing more things to be `unreachable` allows DCE to be more aggressive, but I don't remember if this was an opportunistic change or if doing things the Binaryen IR way caused problems. I'll remove this documentation change from the PR for now and if it was there for a good reason I will find out (again) soon enough.

Reply:
Oh wait, this documentation is not new in this PR; it already existed in stack-utils.h. Would it be ok with you if we revisited this point once things are working end-to-end? I'm concerned that changes we make in this half-built state will come back to bite us before we finish.

Reply:
If it can speed things up I'm not opposed. But I'd like to understand how this can work, though. The point of Poppy IR is to model the stack machine format more closely, and that format doesn't have unreachable blocks. So shouldn't it be the opposite, that is, the example block here would always have type i32 in Poppy IR?

Or am I misunderstanding Poppy IR?

Reply:
I'd say the point of Poppy IR is to make it easy to write optimizations and other transformations that take the stack machine structure of instructions into account. Part of that is necessarily modeling the stack machine syntax more closely, but `unreachable` types still give optimizations useful flexibility and information, so they are worth preserving in Poppy IR.